homeassistant/components/cast/media_player.py

```diff
@@ -488,6 +488,62 @@ def play_media(self, media_type, media_id, **kwargs):
             except NotImplementedError:
                 _LOGGER.error("App %s not supported", app_name)
         else:
+            # ChromeCast will fail to resolve any hostname which is not publicly
+            # resolvable due to it ignoring DHCP supplied DNS server information
+            # and instead using Google's DNS servers.
+            # This is often a problem for HomeAssistant's internal_url, so resolve it
+            # if it's not public and replace with an IP address before passing it to
+            # ChromeCast.
+            # Only resolve if the hostname is not public to avoid breaking vhost
+            # configurations etc.
+            from homeassistant.helpers.network import NoURLAvailableError, get_url
+            from urllib.parse import urlparse, urlunparse
+            import aiodns
+            import socket
+
+            try:
+                internal_url = urlparse(get_url(self.hass, allow_external=False))
+                media_url = urlparse(media_id)
+                if media_url.netloc == internal_url.netloc:
+                    try:
+                        google_dns = ["8.8.8.8", "8.8.4.4"]
+                        response = await aiodns.DNSResolver(google_dns).query(
+                            media_url.hostname, "A"
+                        )
+                    except (aiodns.error.DNSError):
+                        # The hostname is not publicly resolvable, warn and attempt
+                        # local resolution
+                        _LOGGER.info(
+                            "internal_url host %s is not publicly resolvable, attempting to resolve",
+                            internal_url.hostname,
+                        )
+                        try:
+                            response = await aiodns.DNSResolver().gethostbyname(
+                                media_url.hostname, socket.AF_INET
+                            )
```

Comment:
This assumes one is using a local resolver on their HA installation, which doesn't have to be the case (e.g., OS installations use their own resolvers, which may have been failed over)